src/arc.rs

```diff
@@ -183,47 +211,74 @@ impl PathArc {
     /// [`canonicalize`]: struct.PathAbs.html#method.canonicalize
     /// [`current_dir`]: fn.current_dir.html
     pub fn absolute(&self) -> Result<PathAbs> {
-        let mut components = self.components();
-        let mut stack: Vec<OsString> = Vec::new();
-
-        macro_rules! pop_stack { [] => {{
-            if let None = stack.pop() {
-                return Err(Error::new(
-                    io::Error::new(io::ErrorKind::NotFound, ".. consumed root"),
-                    "resolving absolute",
-                    self.clone(),
-                ));
+        let mut res = PathBuf::new();
+
+        fn maybe_init_res(res: &mut PathBuf, resolvee: &PathArc) -> Result<()> {
+            if res.as_os_str() != "" {
+                // res has already been initialized, let's leave it alone.
+                return Ok(());
             }
-        }}}
 
-        handle_prefix(self, &mut stack, &mut components, false)?;
+            // res has not been initialized, let's initialize it to the
+            // canonicalized current directory.
+            let cwd = env::current_dir().map_err(|e| {
+                Error::new(
+                    e,
+                    "getting current_dir while resolving absolute",
+                    resolvee.clone(),
+                )
+            })?;
+            *res = cwd.canonicalize().map_err(|e| {
+                Error::new(e, "canonicalizing", PathArc::new(&cwd))
+            })?;
+
+            Ok(())
+        };
+
+        for each in self.components() {
+            match each {
+                Component::Prefix(p) => {
+                    // We don't care what's already in res, we can entirely
+                    // replace it..
+                    res = make_verbatim_prefix(&p)?;
+                }
 
-        for component in components {
-            match component {
-                Component::CurDir => { /* ignore, probably impossible */ }
-                Component::Prefix(_) => unreachable!(),
                 Component::RootDir => {
-                    if cfg!(unix) {
-                        unreachable!("root is already handled on unix");
+                    if cfg!(windows) {
+                        // A root path component is relative to some existing
+                        // path, so we may need to initialize res.
+                        maybe_init_res(&mut res, self)?;
```

Comment:
won't this set res to `current_dir()`, then try to push onto it?

Are you still relying on the `push` clearing it, or am I missing something here? This code seems way too clear and not explicit enough in what it is trying to do.

Also, I would think that we want to in some way resolve the `each` to be cannonicalized instead of just pushing it onto `res`.

Reply:
> Are you still relying on the `push` clearing it, or am I missing something here?

`push` doesn't entirely clear `res`. For example, if we want to absolutise the path `\foo`:

1. we start with `res` as empty
1. we iterate over `self.components()` and get a `RootDir` component.
1. we call `maybe_init_res()`, setting `res` to the current directory... say, `\\?\C:\Users\vitiral`
1. we push the `RootDir` component to `res`, truncating it to `\\?\C:\`
1. we push the `foo` component to `res`, making it `\\?\C:\foo`

The operation we want here is "truncate to prefix". I'll grant that `res.push(Component::RootDir)` is not the most explicit way to perform that operation, but I can't find a nicer way to say it. It could be `res = res.components.next().unwrap().as_ref::<Path>().to_path_buf()` or possibly a similar thing based on `.ancestors().last()`, but I feel those alternatives are also pretty obscure.

I'll add a comment.

> Also, I would think that we want to in some way resolve the `each` to be cannonicalized instead of just pushing it onto `res`.

My initial implementation was like this; instead of grabbing the current directory and pushing `RootDir`, it just tried to canonicalize RootDir directly (implicitly finding the prefix of the current directory). Unfortunately, [as we discovered](https://github.com/vitiral/path_abs/issues/21#issuecomment-410967402), Windows refuses to canonicalize `\` if the current working directory uses extended-length syntax, so this is our workaround.

I'll put that in the comment too.